primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,142 @@
+use parity_util_mem::MallocSizeOf;
+use polkadot_core_primitives::Hash;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec::Vec};
+
+/// # Execution environment parameter tags
+/// Values from 1 to 16 are reserved for system-wide parameters not related to any concrete
+/// execution environment.
+/// Environment type
+pub const EEPAR_01_ENVIRONMENT: u32 = 1;
+/// Logical stack limit, in stack items
+pub const EEPAR_17_STACK_LOGICAL_MAX: u32 = 17;
+/// Native stack limit, in bytes
+pub const EEPAR_18_STACK_NATIVE_MAX: u32 = 18;
+
+/// # Execution enviroment types
+/// Generic Wasmtime environment
+pub const EXEC_ENV_TYPE_WASMTIME_GENERIC: u32 = 0;
+
+/// Unit type wrapper around [`type@Hash`] that represents an execution parameter set hash.
+///
+/// This type is produced by [`ExecutorParams::hash`].
+#[derive(Clone, Copy, Encode, Decode, Hash, Eq, PartialEq, PartialOrd, Ord, TypeInfo)]
+pub struct ExecutorParamsHash(Hash);
+
+impl sp_std::fmt::Display for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		self.0.fmt(f)
+	}
+}
+
+impl sp_std::fmt::Debug for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "{:?}", self.0)
+	}
+}
+
+impl sp_std::fmt::LowerHex for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		sp_std::fmt::LowerHex::fmt(&self.0, f)
+	}
+}
+
+impl From<Hash> for ExecutorParamsHash {
+	fn from(hash: Hash) -> Self {
+		Self(hash)
+	}
+}
+
+impl From<[u8; 32]> for ExecutorParamsHash {
+	fn from(hash: [u8; 32]) -> Self {
+		Self(hash.into())
+	}
+}
+
+/// # Deterministically serialized execution environment semantics
+/// Represents an arbitrary semantics of an arbitrary execution environment, so should be kept as
+/// abstract as possible. Mapping from `u32` constant tags to SCALE-encoded values was chosen
+/// over an idiomatic `enum` representation as the latter would require caller to rely on
+/// deserialization errors when decoding a structure of newer version not yet known to the node
+/// software, which is undesirable.
+/// As this one goes on-chain, it requires full serialization determinism. That is ensured by
+/// enforcing constant tags to be in ascending order and using deterministic SCALE-codec for values.
+#[cfg_attr(feature = "std", derive(MallocSizeOf))]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub struct ExecutorParams(Vec<(u32, Vec<u8>)>);
+
+impl ExecutorParams {
+	/// Creates a new, empty executor parameter set
+	pub fn new() -> Self {
+		ExecutorParams(Vec::new())
+	}
+
+	/// Returns execution environment type identifier
+	pub fn environment(&self) -> u32 {
+		if self.0.len() < 1 || self.0[0].0 != EEPAR_01_ENVIRONMENT {
+			return EXEC_ENV_TYPE_WASMTIME_GENERIC
+		}
+		let env_enc = self.0[0].1.clone();
+		match u32::decode(&mut &env_enc[..]) {
+			Ok(env) => env,
+			_ => EXEC_ENV_TYPE_WASMTIME_GENERIC,
+		}
```

Comment:
Looks like this is some kind of requirement? Maybe `new()` should take this information.

Reply:
Indeed, not a strict requirement at all. The idea was to make an empty vector a totally valid value producing the default parameter set. So if environment type is there then use it, if it isn't, it only means to use a default environment. It is supposed to be used inside [PVF host's executor interface](https://github.com/paritytech/polkadot/pull/6161/files#diff-f4ffe6a7003b388a382a12c2ac95731f8862f4e3cd96be2898f271c131ee528aR111) in the future, when we have more than one execution environment.

Reply:
Okay, but why not pass always the default env? Or write a `Default` implemenation for the type that uses the default env?

Reply:
[There is a `Default` implementation](https://github.com/paritytech/polkadot/pull/6312/files/3d658041b27cb345fadcde0c20d12ea451b86fe0#diff-59bd1090a32a865b4059c38707d7abe125a65fa65d955abd0df38a1ba6d3c345R138-R142) which returns an empty vector. Indeed, it doesn't make any logical difference neither for current implementation, nor for future ones. If you think that supplying the environment type as a required parameter to `new()` is just a better architecture, I'll make it like that.

cc @pepyakin just to have a second opinion